File: external/bindings/python/mind/tests/test_basic.py
# Copyright 2026 The MIND Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Basic tests for MIND Python bindings.

Can run with pytest or standalone:
    python -m mind.tests.test_basic
"""

import sys
import tempfile
from pathlib import Path


def test_lifecycle():
    """Test basic lifecycle: create, use, destroy."""
    from mind import MindState

    state = MindState(dim=4, slots=8)
    assert state.dim == 4
    assert state.slot_count == 0

    print("PASS: lifecycle")


def test_update_and_query():
    """Test feeding experience and querying."""
    from mind import MindState

    state = MindState(dim=4, slots=8)

    # Feed experience - need multiple updates to build confidence
    # Single update creates slot but confidence=0 (plasticity=1.0, stability=0.0)
    pattern = [1.0, 0.0, 0.0, 0.0]
    for _ in range(10):
        state.update(pattern, delta_t=1.0)

    assert state.slot_count == 1  # Same pattern reinforces, doesn't create new slot

    # Query - now confidence > 0 because plasticity has decayed
    hint = state.query(pattern)
    assert hint.confidence > 0, f"Expected confidence > 0, got {hint.confidence}"
    assert hint.dim == 4

    print("PASS: update_and_query")


def test_plasticity_bounds():
    """Test mercy floor holds."""
    from mind import MindState

    state = MindState(dim=4, slots=8)
    pattern = [1.0, 0.0, 0.0, 0.0]

    # Many updates to try to force plasticity to zero
    for _ in range(1000):
        state.update(pattern, delta_t=1.0)

    p = state.plasticity()
    assert p.plasticity >= 0.05, f"Plasticity {p.plasticity} below mercy floor"
    assert p.plasticity <= 1.0

    print("PASS: plasticity_bounds")


def test_temporal():
    """Test temporal awareness."""
    from mind import MindState

    state = MindState(dim=4, slots=8)
    pattern = [1.0, 0.0, 0.0, 0.0]

    for _ in range(10):
        state.update(pattern, delta_t=1.0)

    t = state.temporal()
    assert t.age == 10.0
    assert t.total_updates == 10
    assert t.maturity > 0

    print("PASS: temporal")


def test_calibration():
    """Test S2S calibration signal."""
    from mind import MindState

    state = MindState(dim=4, slots=8)
    pattern = [1.0, 0.0, 0.0, 0.0]

    for _ in range(50):
        state.update(pattern, delta_t=1.0)

    cal = state.calibration()
    assert cal.age > 0
    assert cal.plasticity > 0
    assert cal.maturity > 0

    print("PASS: calibration")


def test_persistence():
    """Test save and load."""
    from mind import MindState

    with tempfile.TemporaryDirectory() as tmpdir:
        path = Path(tmpdir) / "test.state"

        # Create and save
        state1 = MindState(dim=4, slots=8)
        pattern = [1.0, 0.0, 0.0, 0.0]
        for _ in range(20):
            state1.update(pattern, delta_t=1.0)

        hint1 = state1.query(pattern)
        state1.save(path)

        # Load into new state
        state2 = MindState(dim=4, slots=8)
        state2.load(path)

        hint2 = state2.query(pattern)
        assert hint1.confidence == hint2.confidence, "Persistence failed"

    print("PASS: persistence")


def test_determinism():
    """Test deterministic behavior."""
    from mind import MindState

    patterns = [
        [1.0, 0.0, 0.0, 0.0],
        [0.0, 1.0, 0.0, 0.0],
        [0.5, 0.5, 0.0, 0.0],
    ]

    confidences = []

    for _ in range(2):
        state = MindState(dim=4, slots=16)
        for i in range(100):
            state.update(patterns[i % 3], delta_t=1.0)
        hint = state.query(patterns[0])
        confidences.append(hint.confidence)

    assert confidences[0] == confidences[1], "Non-deterministic behavior"

    print("PASS: determinism")


def test_context_manager():
    """Test context manager interface."""
    from mind import MindState

    with MindState(dim=4, slots=8) as state:
        state.update([1.0, 0.0, 0.0, 0.0], delta_t=1.0)
        assert state.slot_count == 1

    print("PASS: context_manager")


def run_all():
    """Run all tests."""
    print("MIND Python Bindings Test Suite")
    print("================================\n")

    tests = [
        test_lifecycle,
        test_update_and_query,
        test_plasticity_bounds,
        test_temporal,
        test_calibration,
        test_persistence,
        test_determinism,
        test_context_manager,
    ]

    failures = 0
    for test in tests:
        try:
            test()
        except Exception as e:
            print(f"FAIL: {test.__name__}")
            print(f"  {type(e).__name__}: {e}")
            failures += 1

    print("\n================================")
    if failures == 0:
        print("All tests passed.")
        return 0
    else:
        print(f"{failures} test(s) failed.")
        return 1


if __name__ == "__main__":
    sys.exit(run_all())

File: tests/test_basic.c
/*
 * Copyright 2026 The MIND Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * @file test_basic.c
 * @brief Basic tests for MIND
 *
 * These tests verify core invariants.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include "cr.h"

#define ASSERT(cond, msg) do { \
    if (!(cond)) { \
        fprintf(stderr, "FAIL: %s\n  at %s:%d\n", msg, __FILE__, __LINE__); \
        return 1; \
    } \
} while(0)

#define PASS(name) printf("PASS: %s\n", name)

/*============================================================================
 * Test: Lifecycle
 *============================================================================*/

static int test_lifecycle(void) {
    cr_config_t cfg = {4, 8, 1.0f};

    cr_runtime_t* rt = cr_runtime_create(&cfg);
    ASSERT(rt != NULL, "runtime creation");

    cr_state_t* st = cr_state_create(rt);
    ASSERT(st != NULL, "state creation");

    cr_state_destroy(st);
    cr_runtime_destroy(rt);

    PASS("lifecycle");
    return 0;
}

/*============================================================================
 * Test: Plasticity bounds (mercy)
 *============================================================================*/

static int test_plasticity_bounds(void) {
    cr_config_t cfg = {4, 8, 1.0f};
    cr_runtime_t* rt = cr_runtime_create(&cfg);
    cr_state_t* st = cr_state_create(rt);

    float pattern[4] = {1.0f, 0.0f, 0.0f, 0.0f};

    /* Run many updates to try to force plasticity to zero */
    for (int i = 0; i < 10000; i++) {
        cr_state_update(st, pattern, 4, 1.0f);
    }

    cr_plasticity_t p;
    cr_state_plasticity(st, &p);

    /* Plasticity must stay above epsilon (0.05) */
    ASSERT(p.plasticity >= 0.05f, "plasticity floor (mercy)");
    ASSERT(p.plasticity <= 1.0f, "plasticity ceiling");

    cr_state_destroy(st);
    cr_runtime_destroy(rt);

    PASS("plasticity_bounds");
    return 0;
}

/*============================================================================
 * Test: Determinism
 *============================================================================*/

static int test_determinism(void) {
    cr_config_t cfg = {4, 16, 1.0f};

    float patterns[3][4] = {
        {1.0f, 0.0f, 0.0f, 0.0f},
        {0.0f, 1.0f, 0.0f, 0.0f},
        {0.5f, 0.5f, 0.0f, 0.0f}
    };

    /* Run twice with same inputs */
    float confidence1, confidence2;

    for (int run = 0; run < 2; run++) {
        cr_runtime_t* rt = cr_runtime_create(&cfg);
        cr_state_t* st = cr_state_create(rt);

        for (int i = 0; i < 100; i++) {
            cr_state_update(st, patterns[i % 3], 4, 1.0f);
        }

        cr_hint_t hint;
        cr_state_query(st, patterns[0], 4, &hint);

        if (run == 0) {
            confidence1 = hint.confidence;
        } else {
            confidence2 = hint.confidence;
        }

        cr_state_destroy(st);
        cr_runtime_destroy(rt);
    }

    /* Must be exactly equal (determinism) */
    ASSERT(confidence1 == confidence2, "deterministic results");

    PASS("determinism");
    return 0;
}

/*============================================================================
 * Test: Bounded memory
 *============================================================================*/

static int test_bounded_memory(void) {
    cr_config_t cfg = {4, 4, 1.0f};  /* Only 4 slots */
    cr_runtime_t* rt = cr_runtime_create(&cfg);
    cr_state_t* st = cr_state_create(rt);

    /* Try to add 10 different patterns */
    for (int i = 0; i < 10; i++) {
        float pattern[4] = {
            (float)i,
            (float)(i * 2),
            (float)(i * 3),
            (float)(i * 4)
        };
        cr_state_update(st, pattern, 4, 1.0f);
    }

    /* Slot count must not exceed max */
    int slots = cr_state_slot_count(st);
    ASSERT(slots <= 4, "bounded slot count");

    cr_state_destroy(st);
    cr_runtime_destroy(rt);

    PASS("bounded_memory");
    return 0;
}

/*============================================================================
 * Test: Age is monotonic
 *============================================================================*/

static int test_age_monotonic(void) {
    cr_config_t cfg = {4, 8, 1.0f};
    cr_runtime_t* rt = cr_runtime_create(&cfg);
    cr_state_t* st = cr_state_create(rt);

    float pattern[4] = {1.0f, 0.0f, 0.0f, 0.0f};
    float prev_age = 0.0f;

    for (int i = 0; i < 100; i++) {
        cr_state_update(st, pattern, 4, 0.5f);

        cr_temporal_t t;
        cr_state_temporal(st, &t);

        ASSERT(t.age > prev_age, "age increases");
        prev_age = t.age;
    }

    cr_state_destroy(st);
    cr_runtime_destroy(rt);

    PASS("age_monotonic");
    return 0;
}

/*============================================================================
 * Test: Persistence
 *============================================================================*/

static int test_persistence(void) {
    cr_config_t cfg = {4, 8, 1.0f};
    const char* path = "/tmp/mind_test.state";

    float pattern[4] = {1.0f, 0.0f, 0.0f, 0.0f};
    float saved_confidence, loaded_confidence;

    /* Create, update, save */
    {
        cr_runtime_t* rt = cr_runtime_create(&cfg);
        cr_state_t* st = cr_state_create(rt);

        for (int i = 0; i < 50; i++) {
            cr_state_update(st, pattern, 4, 1.0f);
        }

        cr_hint_t hint;
        cr_state_query(st, pattern, 4, &hint);
        saved_confidence = hint.confidence;

        int err = cr_state_save(st, path);
        ASSERT(err == 0, "save succeeds");

        cr_state_destroy(st);
        cr_runtime_destroy(rt);
    }

    /* Load, query */
    {
        cr_runtime_t* rt = cr_runtime_create(&cfg);
        cr_state_t* st = cr_state_create(rt);

        int err = cr_state_load(st, path);
        ASSERT(err == 0, "load succeeds");

        cr_hint_t hint;
        cr_state_query(st, pattern, 4, &hint);
        loaded_confidence = hint.confidence;

        cr_state_destroy(st);
        cr_runtime_destroy(rt);
    }

    ASSERT(saved_confidence == loaded_confidence, "persistence preserves state");

    PASS("persistence");
    return 0;
}

/*============================================================================
 * Test: S2S calibration signal
 *============================================================================*/

static int test_calibration(void) {
    cr_config_t cfg = {4, 8, 1.0f};
    cr_runtime_t* rt = cr_runtime_create(&cfg);
    cr_state_t* st = cr_state_create(rt);

    float pattern[4] = {1.0f, 0.0f, 0.0f, 0.0f};

    for (int i = 0; i < 100; i++) {
        cr_state_update(st, pattern, 4, 1.0f);
    }

    cr_calibration_t cal;
    int err = cr_state_calibration(st, &cal);

    ASSERT(err == 0, "calibration export");
    ASSERT(cal.age > 0.0f, "calibration has age");
    ASSERT(cal.plasticity > 0.0f, "calibration has plasticity");
    ASSERT(cal.maturity > 0.0f, "calibration has maturity");

    cr_state_destroy(st);
    cr_runtime_destroy(rt);

    PASS("calibration");
    return 0;
}

/*============================================================================
 * Main
 *============================================================================*/

int main(void) {
    printf("MIND Test Suite\n");
    printf("================\n\n");

    int failures = 0;

    failures += test_lifecycle();
    failures += test_plasticity_bounds();
    failures += test_determinism();
    failures += test_bounded_memory();
    failures += test_age_monotonic();
    failures += test_persistence();
    failures += test_calibration();

    printf("\n================\n");
    if (failures == 0) {
        printf("All tests passed.\n");
        return 0;
    } else {
        printf("%d test(s) failed.\n", failures);
        return 1;
    }
}
